### DISCUSSION TITLE
A little more understanding of the signatures in PDF

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#105] [General]
A simple understanding of how this stuff works.
https://pspdfkit.com/blog/2018/digital-signatures/

Why doesn't pyHanko not require private key for validation?

**Comment by MatthiasValvekens** [comment]
Well, you have it exactly backwards: the private key is used to create signatures, not to validate. If the public key was sufficient to create signatures, then everyone could forge signatures by everyone else, so what would even be the point? ;)
Conversely you typically want (basically) anyone to be able to validate your signatures, so it wouldn't make sense to have that require the private key, would it?

Rule of thumb: you use the _private_ key to create a signature (since only the holder of the private key is supposed to be able to create valid signatures), and others use the corresponding _public_ key to validate it.

Besides, the author of that blog post is guilty of an unfortunately very common equivocation that I _really_ dislike, which is equating signing/verifying to decryption/encryption of hash values. That's mathematically kind of true in RSA if you squint hard enough, but (a) it makes zero sense from a security analysis perspective (the security requirements of these procedures are _vastly_ different), and (b) that analogy breaks down spectacularly once you start considering non-RSA signing schemes like DSA, ECDSA and EdDSA. In none of those digital signature schemes, the validation procedure involves anything resembling "decrypting a hash". In fact, none of those cryptosystems even have a related encryption primitive that operates in the same way as RSA's.

So don't put too much stock in that blog post (it's basically a marketing/devrel piece), and just remember that signing requires a private key, while verifying requires the corresponding public one.

**Reply by Vishwamithra37** [reply]
Hey, Thanks for the reply :). I have one more question, how do I get the I printed inside the signature field? Like during the signing, I put in a bunch of stuff visible to the reader. 
Is it possible to retrieve it during validation since it is in the signature field? Is another library required to see the URL in QR code+other information I put in a square signature field? etc.

**Reply by MatthiasValvekens** [reply]
Yep, that's certainly possible. PyHanko has `QRStampStyle` and `TextStampStyle` available for those purposes.
Those only do very basic things, though. Documentation:

 -  [Text-based stamps](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#text-based-stamps)
 -  [Stamps with QR codes](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#qr-code-stamps)

If you need something more fancy/advanced, you can [generate the appearance using external tools as well](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#static-content-stamps), but that's more complicated.

Note that a visible signature field (with an actual presence on the page) is required in all cases, but the code samples show that as well.

**Reply by Vishwamithra37** [reply]
Sorry for bothering, I am able to generate the field. As you pointed out. WIth my custom data.

I want to retrieve the QR code/URL during validation from the PDF. Without actually seeing PDF, but only through pyHanko code. 
Like, I nake a stamp, but now I want to view the stamp details during validation.

**Reply by MatthiasValvekens** [reply]
Ah, sorry, I misunderstood your question. The short answer is no. The PDF standard doesn't provide a mechanism to "attach" a URL to a signature in an easily retrievable way.

There are several ways you could proceed:

 - If you know you'll only ever need to validate signatures provided by your software, then you can always write the URL to some private object in the PDF file prior to computing the document digest, and read it back from there when validating.
 - If you need to extract these URLs from signatures provided by others, then you have a much harder problem on your hands. You'll need a library that is capable of extracting text from annotations (that's not quite as trivial as it sounds).
- If you're happy just processing the QR code, rendering the PDF file to an image (including all annotations) and retrieving the URLs using a QR code scanning library is likely the easiest. Ugly, but it works.
